src/mir/index.md

```diff
@@ -255,7 +255,66 @@ but [you can read about those below](#promoted)).
 
 ## Representing constants
 
-*to be written*
+When code has reached the MIR stage, constants can generally come in two forms: *MIR constants* ([`mir::Constant`]) and *type system constants* ([`ty::Const`]).
+MIR constants are used as operands: in `x + CONST`, `CONST` is a MIR constant; similarly, in `x + 2`, `2` is a MIR constant.
+Type system constants are used in the type system, in particular for array lengths but also for const generics.
+
+Generally, both kinds of constants can be "unevaluated" or "already evaluated".
+And unevaluated constant simply stores the `DefId` of what needs to be evaluated to compute this result.
+An evaluated constant (a "value") has already been computed; their representation differs between type system constants and MIR constants:
+MIR constants evaluate to a `mir::ConstValue`; type system constants evaluate to a `ty::ValTree`.
+
+Type system constants have some more variants to support const generics: they can refer to local const generic parameters, and they are subject to inference.
+Furthermore, the `mir::Constant::Ty` variant lets us use an arbitrary type system constant as a MIR constant; this happens whenever a const generic parameter is used as an operand.
+
+### MIR constant values
+
+In general, a MIR constant value (`mir::ConstValue`) was computed by evaluating some constant the user wrote.
+This [const evaluation](../const-eval.md) produces a very low-level representation of the result in terms of individual bytes.
+We call this an "indirect" constant (`mir::ConstValue::Indirect`) since the value is stored in-memory.
+
+However, storing everything in-memory would be awfully inefficient. Hence there
+are some other variants in `mir::ConstValue` that can represent certain simple
+and common values more efficiently. In particular, everything that can be
+directly written as a literal in Rust (integers, floats, chars, bools, but also
+`"string literals"` and `b"byte string literals"`) has an optimized variant that
+avoids the full overhead of the in-memory representation.
+
+### ValTrees
```

Comment:
This is arguably a bit misplaced in the MIR section, but I didn't find a "type system" section. previously it was in the const-eval section and that seems worse.